### ISSUE TITLE
http_build_query

### ISSUE THREAD

**Issue by robertmain** [robertmain/communique#43]
`http_build_query` is expecting a non-scalar data type. JSON encoding data prior to being sent over the  wire breaks this.

**Comment by robertmain** [comment]
We should probably throw an exception if we pass http_build_query a ~~non-scalar~~ scalar value.

**Comment by StewartAtkins** [comment]
Or add appropriate type checking so that you don't try to build a querystring with something that isn't an array (or other compatible object)

**Comment by robertmain** [comment]
See previous comment

**Comment by StewartAtkins** [comment]
Your previous comment makes little sense on two fronts:
1. You suggest throwing an exception in the case of a non-scalar value, but non-scalar values are what you WANT to be passing to http_build_query
2. I said nothing about throwing an exception, only type checking. You also WANT to be able to accept scalar values, you just don't want to run them through http_build_query first. This will allow you to continue to support payload-transforming interceptors like the JSON interceptor and also allow users to pass a plain string in the case that they want to set a raw POST body.

**Comment by robertmain** [comment]
```php
array (
'date' => new Date()
);
```

Try passing that to `http_build_query_ then tell me I shouldn't throw an exception.

**Comment by StewartAtkins** [comment]
`Fatal error: Class 'Date' not found in Command line code on line 1`

Conversely...
```
array (
'date' => new DateTimeZone("UTC")
);
```
yields the perfectly acceptable
`string(42) "date%5Btimezone_type%5D=3&date%5Btimezone%5D=UTC"`
but well what about something weirder?

```
array (
'x' => new ZipArchive()
);
```
yields the also perfectly acceptable
`string(85) "x%5Bstatus%5D=0&x%5BstatusSys%5D=0&x%5BnumFiles%5D=0&x%5Bfilename%5D=&x%5Bcomment%5D="`

I'm not going to argue that these are the most sensible representations in the world but they don't appear to represent any sort of error condition, why would they imply that an exception should be raised? Perhaps if we look at a more realistic example:
```
$x = json_decode('{\"a\": \"b\", \"c\": {\"d\": \"e\", \"f\": [1,2,3,\"g\"]}}');
var_dump($x);
```
```object(stdClass)#1 (2) {
  ["a"]=>
  string(1) "b"
  ["c"]=>
  object(stdClass)#2 (2) {
    ["d"]=>
    string(1) "e"
    ["f"]=>
    array(4) {
      [0]=>
      int(1)
      [1]=>
      int(2)
      [2]=>
      int(3)
      [3]=>
      string(1) "g"
    }
  }
}
```
so that also results in an object rather than an array, but http_build_query seems fine with it:
```
var_dump(http_build_query(array("x" => $x)));
string(128) "x%5Ba%5D=b&x%5Bc%5D%5Bd%5D=e&x%5Bc%5D%5Bf%5D%5B0%5D=1&x%5Bc%5D%5Bf%5D%5B1%5D=2&x%5Bc%5D%5Bf%5D%5B2%5D=3&x%5Bc%5D%5Bf%5D%5B3%5D=g"
```

Interestingly the only case I've found that doesn't seem to work quite as one might expect is if the class has a __toString method in which case http_build_query doesn't call it, it simply serialises the properties as in the above examples, but it still doesn't represent an error condition.

**Comment by robertmain** [comment]
hmm, that is odd, my (brief) tests I found that `http_build_query` didn't like being passed objects/arrays etc. as the value of the array. Might have to look into that further.

**Comment by robertmain** [comment]
Yeah, basically. PHP's [`is_scalar()`](http://php.net/manual/en/function.is-scalar.php) should take care of this.
